File: god-sim-prototype/src/rng.js
/* SLATE Phase 0 — deterministic seeded RNG streams.
   Every system pulls from its own labeled stream so systems can be
   added/removed without reshuffling everyone else's randomness. */
(function (S) {
  'use strict';

  function hashStr(s) {
    let h = 2166136261 >>> 0;
    for (let i = 0; i < s.length; i++) {
      h ^= s.charCodeAt(i);
      h = Math.imul(h, 16777619);
    }
    return h >>> 0;
  }

  function mulberry32(a) {
    return function () {
      a |= 0; a = (a + 0x6D2B79F5) | 0;
      let t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  function makeRng(seed, label) {
    const fn = mulberry32(hashStr(String(seed) + '/' + label));
    return {
      next: fn,
      range: (a, b) => a + fn() * (b - a),
      int: (a, b) => a + Math.floor(fn() * (b - a + 1)),
      chance: (p) => fn() < p,
      pick: (arr) => arr[Math.floor(fn() * arr.length)],
      shuffle: (arr) => {
        const a = arr.slice();
        for (let i = a.length - 1; i > 0; i--) {
          const j = Math.floor(fn() * (i + 1));
          [a[i], a[j]] = [a[j], a[i]];
        }
        return a;
      },
    };
  }

  // Stateless position hash in [0,1) — for per-cell visual jitter.
  function hash2(x, y, seed) {
    let n = Math.imul(x, 374761393) ^ Math.imul(y, 668265263) ^ (seed | 0);
    n = Math.imul(n ^ (n >>> 13), 1274126177);
    return ((n ^ (n >>> 16)) >>> 0) / 4294967296;
  }

  S.rng = { makeRng, hashStr, mulberry32, hash2 };
})(typeof SLATE !== 'undefined' ? SLATE : (globalThis.SLATE = {}));

File: SlateSim/Runtime/Rng.cs
// SLATE — deterministic seeded RNG streams. Port of god-sim-prototype/src/rng.js.
// Every system pulls from its own labeled stream so systems can be added or
// removed without reshuffling everyone else's randomness. All integer math is
// wrapped 32-bit to match the JS prototype bit-for-bit.
using System;
using System.Collections.Generic;

namespace Slate.Sim
{
    public static class SlateRng
    {
        // FNV-1a, identical to the JS hashStr (Math.imul == wrapped 32-bit multiply).
        public static uint HashStr(string s)
        {
            unchecked
            {
                uint h = 2166136261u;
                for (int i = 0; i < s.Length; i++)
                {
                    h ^= s[i];
                    h *= 16777619u;
                }
                return h;
            }
        }

        // Stateless position hash in [0,1) — for per-cell jitter.
        public static double Hash2(int x, int y, uint seed)
        {
            unchecked
            {
                int n = (x * 374761393) ^ (y * 668265263) ^ (int)seed;
                n = (n ^ (int)((uint)n >> 13)) * 1274126177;
                uint r = (uint)(n ^ (int)((uint)n >> 16));
                return r / 4294967296.0;
            }
        }
    }

    // mulberry32 stream, keyed "<seed>/<label>".
    public sealed class Rng
    {
        private int _a;

        public Rng(int seed, string label)
        {
            _a = unchecked((int)SlateRng.HashStr(seed.ToString() + "/" + label));
        }

        public double Next()
        {
            unchecked
            {
                _a += 0x6D2B79F5;
                int t = (_a ^ (int)((uint)_a >> 15)) * (1 | _a);
                t = (t + ((t ^ (int)((uint)t >> 7)) * (61 | t))) ^ t;
                uint r = (uint)(t ^ (int)((uint)t >> 14));
                return r / 4294967296.0;
            }
        }

        public double Range(double a, double b) => a + Next() * (b - a);

        public int Int(int a, int b) => a + (int)Math.Floor(Next() * (b - a + 1));

        public bool Chance(double p) => Next() < p;

        public T Pick<T>(IReadOnlyList<T> list) => list[(int)Math.Floor(Next() * list.Count)];

        public List<T> Shuffle<T>(IReadOnlyList<T> list)
        {
            var a = new List<T>(list);
            for (int i = a.Count - 1; i > 0; i--)
            {
                int j = (int)Math.Floor(Next() * (i + 1));
                (a[i], a[j]) = (a[j], a[i]);
            }
            return a;
        }
    }
}
